torch/distributed/fsdp/flatten_params_wrapper.py

```diff
@@ -70,407 +57,98 @@ def _pre_load_state_dict_hook(
-        # `getattr(module, "flat_param") is None`. This is not always accurate
-        # as the above condition is also true if this module is unflattened.
-        # `no_params` explicitly shows this module has no parameters and
-        # is always correct regardless flattened or unflattened.
-        self.no_params = True
         self.flat_param = None
-
-        # Register hook to be called after state_dict() to remove the
-        # "_fpw_module." prefix and before load_state_dict() to add it back.
-        # The hooks must be registered even if the target param_list is empty as
-        # all submodules in FlattenParamsWrapper should be pre/post processed by
-        # the hooks.
+        # Register hooks to clean parameter names for state dict (even if this
+        # wrapper itself manages no parameters since it must clean names from
+        # submodules)
         self._register_state_dict_hook(_post_state_dict_hook)
         self._register_load_state_dict_pre_hook(_pre_load_state_dict_hook)
-
-        if len(param_list) == 0:
+        if len(params) == 0:
             return
+        self._flat_param_handle = FlatParamHandle(params, module)
+        # Defining `self.flat_param` registers the `FlatParameter` and makes it
+        # visible to `named_parameters()`
+        self.flat_param = self._flat_param_handle.flat_param
 
-        # A list of parameters to be flatten
-        unique_param_list = set(param_list)
-        self.no_params = False
-
-        # convert from list of Parameters to set of (Module, parameter_name) tuples, which
-        # will survive in case the Parameter instances are reset.
-        # it includes (m, n) that points to the same parameter.
-        self.param_set = set()
-        for m in self.modules():
-            for n, p in m.named_parameters(recurse=False):
-                if p in unique_param_list:
-                    self.param_set.add((m, n))
-
-        params, param_infos, shared_param_infos = self._init_flatten_params()
-        self.flat_param = FlatParameter(params, params[0].requires_grad)
-        self.flat_param._param_infos = param_infos
-        self.flat_param._shared_param_infos = shared_param_infos
-
-        # This attribute is used to remember the flat_param inside the unflatten_params()
-        # context. With this attribute, FSDP can access the flat parameter metadata
-        # even if flat_param is temporarily deleted.
-        # ``orig_flat_param` is a list to avoid being tracked by ``state_dict()``.
-        self.orig_flat_param: List[Optional[FlatParameter]] = [None]
-        self._flatten_params()
+    @property
+    def has_params(self) -> bool:
+        """Returns whether this wrapper manages any parameters."""
+        return hasattr(self, "_flat_param_handle")
 
-        # Sanity check for the string constants.
-        assert getattr(self, FPW_MODULE) is self._fpw_module
-        assert getattr(self, FLAT_PARAM) is self.flat_param
```

Comment:
cc @fegin whether we need to keep the two asserts, I remembered it is intentionally added?

Reply:
They are intentionally added to ensure FPW_MODULE and FLAT_PARAM match the actually naming in the module. It won't affect any functions if the naming match.

Reply:
I added the `assert`s back in :)